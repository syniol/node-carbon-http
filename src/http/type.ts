import {
  ClientRequest,
  IncomingMessage as InSecureIncomingMessage,
  RequestOptions as InSecureRequestOpt,
} from 'http'
import { RequestOptions as SecureRequestOpt } from 'https'
import { URL } from 'url'

import { HttpStatusCode } from './codes'

export type RequestType = (
  options: InSecureRequestOpt | SecureRequestOpt | string | URL,
  callback?: (res: InSecureIncomingMessage) => void,
) => ClientRequest

export enum HttpMethod {
  GET = 'GET',
  POST = 'POST',
  PUT = 'PUT',
  DELETE = 'DELETE',
  PATCH = 'PATCH',
}

export interface HttpRequestContext {
  [key: string]: any

  headers?: any
  method?: HttpMethod
  body?: string
  port?: number
}

export interface HttpResponse {
  status: HttpStatusCode

  text(): string
  json(): any
}
