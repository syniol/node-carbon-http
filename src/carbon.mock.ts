import { HttpStatusCode, NodeRequestClient } from './http'
import { Buffer } from 'buffer'
import { ClientRequest, IncomingMessage as InSecureIncomingMessage } from 'http'


export default class CarbonHTTPMock {
  public readonly client: NodeRequestClient

  public constructor(arg: string | Error, statusCode?: HttpStatusCode) {
    this.client = (_, callback) => {
      if (typeof callback === 'function') {
        const res = (<unknown>{
          on: (e: 'data' | 'error' | 'end', listener: (data?: any) => {}) => {
            switch (e) {
              case 'data':
                if (typeof arg === 'string') {
                  listener(Buffer.from(arg, 'utf8'))
                }
                break
              case 'error':
                if (arg instanceof Error) {
                  listener(arg)
                }
                break
              default:
                listener()
            }

            return this
          },
          statusCode:
            statusCode ||
            (
              arg instanceof Error
              ? HttpStatusCode.BAD_REQUEST
              : HttpStatusCode.OK
            ),
        }) as InSecureIncomingMessage

        callback(res)
      }

      return (<unknown>{
        write: () => true,
        end: () => undefined,
      }) as ClientRequest
    }
  }
}
