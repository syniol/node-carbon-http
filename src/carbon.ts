
import { URL } from 'url'
import { TextEncoder } from 'util'
import { request as InSecureRequest } from 'http'
import { request as SecureRequest } from 'https'

import {
  NodeRequestClient,
  NodeRequestOption,
  CarbonHttpRequestOption,
  CarbonHttpResponse,
  HttpStatusCode,
  HttpMethod,
  HttpProtocol,
} from './http'


function response(
  dataBlocks: Uint8Array[],
  status: Readonly<number>
): Readonly<CarbonHttpResponse> {
  const result: Readonly<string> = Buffer
    .concat(dataBlocks)
    .toString()

  return {
    status: status,
    json(): any {
      return JSON.parse(result)
    },
    text(): string {
      return result
    },
  }
}

export function Request(
  url: Readonly<string>,
  opt?: CarbonHttpRequestOption,
  clientService?: NodeRequestClient | any,
): Promise<Readonly<CarbonHttpResponse>> {
  const urlAPI = new URL(url)
  const nodeReqOpt: NodeRequestOption = {
    method: opt?.method || HttpMethod.GET,
    path: `${urlAPI.pathname}${urlAPI.search || ''}`,
    body: new TextEncoder().encode(opt?.body),
    port: opt?.port ? opt.port : urlAPI.port,
  }

  if (opt?.headers) {
    nodeReqOpt.headers = opt.headers;
  }

  let client = InSecureRequest;
  if (clientService) {
    client = clientService;
  }

  if (urlAPI.protocol === HttpProtocol.SecureHTTP) {
    if (!clientService) {
      client = SecureRequest
    }

    nodeReqOpt.hostname = urlAPI.hostname;
  } else {
    nodeReqOpt.host = urlAPI.host;
  }

  return new Promise((resolve, reject) => {
    const req = client(nodeReqOpt, (res) => {
      const dataCollection: Uint8Array[] = []
      res.on('data', (data: Uint8Array) => {
        dataCollection.push(data)
      })

      res.on('error', (err: any) => {
        reject(err)
      })

      res.on('end', () => {
        resolve(
          response(
            dataCollection,
            res.statusCode || HttpStatusCode.OK
          ),
        )
      })
    })

    req.write(opt?.body || '')

    req.end()
  })
}
