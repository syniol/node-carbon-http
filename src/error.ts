export interface CarbonError extends Error {
  code?: number
}

export function NewCarbonError(
  msg: string,
  name?: string,
  stack?: string,
  code?: number,
): Readonly<CarbonError> {
  const err: CarbonError = {
    name: name || 'CarbonHttpException',
    message: msg,
  }

  if (stack) {
    err.stack = stack
  }

  if (code) {
    err.code = code
  }

  return err
}
