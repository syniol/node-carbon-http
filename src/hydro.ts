
import { URL } from 'url'
import { TextEncoder } from 'util'
import { request as InSecureRequest } from 'http'
import { request as SecureRequest } from 'https'

import {
  RequestType,
  HttpMethod,
  HttpRequestContext,
  HttpResponse,
  HttpStatusCode
} from './http'


export default class HydroHTTP {
  readonly #forcedClient: boolean;
  #client: RequestType

  public constructor(client?: RequestType) {
    this.#forcedClient = false;
    this.#client = InSecureRequest

    if (client) {
      this.#client = client
      this.#forcedClient = true;
    }
  }

  private response(
    dataBlocks: Uint8Array[],
    status: number
  ): Readonly<HttpResponse> {
    const result = Buffer.concat(dataBlocks).toString()

    return {
      status: status,
      json(): any {
        return JSON.parse(result)
      },
      text(): string {
        return result
      },
    }
  }

  public request(
    url: string,
    context?: HttpRequestContext,
  ): Promise<HttpResponse> {
    const urlObject = new URL(url)
    const opt = {
      method: context?.method || HttpMethod.GET,
      headers: context?.headers,
      path: `${urlObject.pathname}${urlObject.search || ''}`,
      body: new TextEncoder().encode(context?.body),
      port: context?.port ? context?.port : urlObject.port,
    }

    if (urlObject?.protocol === 'https:') {
      if (!this.#forcedClient) {
        this.#client = SecureRequest
      }

      Object.assign(opt, { hostname: urlObject.hostname })
    } else {
      Object.assign(opt, { host: urlObject.host })
    }

    return new Promise((resolve, reject) => {
      const req = this.#client(opt, (res) => {
        const dataCollection: Uint8Array[] = []
        res.on('data', (data: Uint8Array) => {
          dataCollection.push(data)
        })

        res.on('error', (err: any) => {
          reject(err)
        })

        res.on('end', () => {
          resolve(
            this.response(
              dataCollection,
              res?.statusCode || HttpStatusCode.OK
            ),
          )
        })
      })

      req.write(context?.body || '')

      req.end()
    })
  }
}
