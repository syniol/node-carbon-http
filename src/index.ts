export * from './http'
export * from './carbon'
export * from './error'
