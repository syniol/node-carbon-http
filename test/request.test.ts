import HydroHTTP from '../src/hydro'
import { HttpResponse } from '../src/http'

describe('Request Test', () => {
  let sut: HydroHTTP

  beforeAll(() => {
    sut = new HydroHTTP()
  })

  it('should be instantiable', () => {
    expect(sut).toBeInstanceOf(HydroHTTP)
  })

  describe('given GET request is made', () => {
    let actual: HttpResponse;

    beforeAll(async () => {
      actual = await sut.request('https://api.genderize.io/?name=hadi')
    })

    it('should bring a response back in JSON format', () => {
      expect(actual.json()).toMatchSnapshot({
        probability: expect.any(Number),
        count: expect.any(Number),
      })
    })

    it('should bring a response back in String format', () => {
      expect(actual.text()).toContain('"name":"hadi","gender":"male"')
    })
  })
})
