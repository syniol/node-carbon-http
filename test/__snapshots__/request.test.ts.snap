// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request Test given GET request is made should bring a response back in JSON format 1`] = `
Object {
  "count": Any<Number>,
  "gender": "male",
  "name": "hadi",
  "probability": Any<Number>,
}
`;
